File: 2022/day02/day02.nim
import strutils, sequtils
# 0 = rock, 1=paper, 2=scissors
const
    A = ord('A')
    X = ord('X')

# returns the move I need to do in order to win m
proc win(m: int): int = (m+1) mod 3

# returns the move I need to do in order to lose m
proc lose(m: int): int =
    result = 2
    if m > 0: result = m-1

proc score(other, me: int): int =
    result = me+1 # base value is my move
    if other == me:
        result+=3 # draw match
    elif win(other) == me:
        result+=6 # won match

proc part1(pl: seq[char]): int =
    score(int(pl[0])-A, int(pl[1])-X)

# X = lose, Y=draw, Z=win
proc part2(pl: seq[char]): int =
    let other = int(pl[0])-A
    case pl[1]:
        of 'X': return score(other, other.lose)
        of 'Y': return score(other, other)
        of 'Z': return score(other, other.win)
        else: discard

var total1, total2: int
for line in "input.txt".lines:
    let players = line.split.mapIt(char(it[0]))
    total1+=players.part1
    total2+=players.part2

echo "part1=", total1
echo "part2=", total2

File: 2022/day02/day02.rb

SYMBOL_TABLE={'A'=> :rock,'B'=>:paper, 'C'=> :scissors, 'X'=>:rock,'Y'=>:paper,'Z'=>:scissors }
SCORE_TABLE={:rock => 1,:paper=>2, :scissors=>3}
LOSE_TABLE={:rock => :scissors, :paper=> :rock, :scissors=>:paper}
WIN_TABLE={:rock=>:paper, :paper=>:scissors, :scissors=>:rock}

def score(other,me)
    return 3+SCORE_TABLE[me] if other==me
    result=SCORE_TABLE[me]
    result+=6 if WIN_TABLE[other]==me
    result
end

def part1(p1,p2)
    score(SYMBOL_TABLE[p1],SYMBOL_TABLE[p2])
end

def part2(p1,p2)
    other=SYMBOL_TABLE[p1]
    # X = lose, Y=draw, Z=win 
    case p2
        when 'X' then score(other,LOSE_TABLE[other])
        when 'Y' then score(other,other) # play same move
        when 'Z' then score(other,WIN_TABLE[other])
    end
end

total1, total2=0,0
File.read("input.txt").split("\n").each do |line|
    p1,p2=line.split 
    total1+=part1(p1,p2)
    total2+=part2(p1,p2)
end
puts total1,total2